### PR TITLE
distsql: let the coprocessor streaming API use StreamResponse

### PR DESCRIPTION
**Pull request by tiancaiamao**
SelectResponse will be replaced by StreamResponse in the future.
This PR assert that the response returned by streaming API would be `StreamResponse`, but not in arrow format yet.

### CODE REVIEW

**Comment by coocood** [comment]
I can't find `StreamResponse` in this PR.

**Comment by tiancaiamao** [comment]
It's already in the vendor https://github.com/pingcap/tidb/pull/5367/files#diff-b82980ae0fe15a500680eceb6db0f845R412
@coocood

**Comment by tiancaiamao** [comment]
@hicqu We can test this when you're ready.

**Comment by tiancaiamao** [comment]
Test with TiKV, it should be OK now.
PTAL @hicqu @coocood  @shenli

**Comment by tiancaiamao** [comment]
@zz-jason @zimulala

File: distsql/distsql.go [hunk 1]
```diff
@@ -234,8 +234,12 @@ func (pr *partialResult) Next(goCtx goctx.Context) (data []types.Datum, err erro
 	if chunk == nil {
 		return nil, nil
 	}
-	data = make([]types.Datum, pr.rowLen)
-	for i := 0; i < pr.rowLen; i++ {
+	return readRowFromChunk(chunk, pr.rowLen)
+}
+
+func readRowFromChunk(chunk *tipb.Chunk, rowLen int) (data []types.Datum, err error) {
```
Comment (zz-jason) [inline distsql/distsql.go]: s/rowLen/numCols/

File: distsql/stream.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+		err = r.readRowsData(chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// readDefaultFromResponse read the data to chunk. Returns true means the resp is finished.
+func readChunkFromResponse(resp kv.Response, chunk *tipb.Chunk) (bool, error) {
```
Comment (zimulala) [inline distsql/stream.go]: `tipb.Chunk` and `chunk.Chunk` are hard to read.

File: distsql/stream.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+	err = chunk.Unmarshal(stream.Data)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return false, nil
+}
+
+// takeData ensure there're data is in current chunk. if no more data, return false.
+func (r *streamResult) takeData() error {
```
Comment (zz-jason) [inline distsql/stream.go]: what is this function used for ?

File: distsql/stream.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	if finish {
+		r.chunk = nil
+		return nil
+	}
+	r.chunk = tmp
+	return nil
+}
+
+func (r *streamResult) readRowsData(chk *chunk.Chunk) (err error) {
```
Comment (zz-jason) [inline distsql/stream.go]: s/readRowsData/flushToChunk/

File: distsql/stream.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+// streamResult implements the SelectResult interface.
+type streamResult struct {
+	resp       kv.Response
+	rowLen     int
+	fieldTypes []*types.FieldType
+	ctx        context.Context
+
+	// NOTE: chunk == nil means finish, while len(chunk.RowsData) == 0 doesn't.
+	chunk *tipb.Chunk
```
Comment (zimulala) [inline distsql/stream.go]: Can we rename it? rawData、copData or others

File: distsql/stream.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	return nil
+}
+
+func (r *streamResult) readRowsData(chk *chunk.Chunk) (err error) {
+	rowsData := r.chunk.RowsData
+	maxChunkSize := r.ctx.GetSessionVars().MaxChunkSize
+	timeZone := r.ctx.GetSessionVars().GetTimeZone()
+	for chk.NumRows() < maxChunkSize && len(rowsData) > 0 {
+		for i := 0; i < r.rowLen; i++ {
+			rowsData, err = codec.DecodeOneToChunk(rowsData, chk, i, r.fieldTypes[i], timeZone)
```
Comment (zimulala) [inline distsql/stream.go]: Maybe remainRowsData is better?

File: distsql/stream.go (hunk 4 above)
Comment (zimulala) [inline distsql/stream.go]: Is this function used to decode rows data?

File: distsql/stream.go (hunk 2 above)
Comment (tiancaiamao) [reply distsql/stream.go]: In the future tipb.Chunk will be removed....

[the remark this answers is not in the document]
Comment (tiancaiamao) [reply distsql/stream.go]: 1. It doesn't use any member in `streamResult`, so why we need make it a  method
2. `streamResult` has `takeData` method, basicly do the same thing.

File: distsql/stream.go (hunk 3 above)
Comment (tiancaiamao) [reply distsql/stream.go]: The comment says:

> // takeData ensure there're data is in current chunk. if no more data, return false.

NextChunk will use this function.

File: distsql/stream.go (hunk 5 above)
Comment (tiancaiamao) [reply distsql/stream.go]: No, it's in protobuf....
One day we'll get rid of it, maybe.

File: distsql/stream.go (hunk 4 above)
Comment (tiancaiamao) [reply distsql/stream.go]: This function read many rows, convert them to chunk.Chunk.

File: distsql/stream.go (hunk 5 above)
Comment (zimulala) [reply distsql/stream.go]: I mean the field of `chunk`.

**Comment by tiancaiamao** [comment]
@zimulala @zz-jason 
PTAL, thanks.

File: distsql/stream.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+	timeZone := r.ctx.GetSessionVars().GetTimeZone()
+	for chk.NumRows() < maxChunkSize && len(remainRowsData) > 0 {
+		for i := 0; i < r.rowLen; i++ {
+			remainRowsData, err = codec.DecodeOneToChunk(remainRowsData, chk, i, r.fieldTypes[i], timeZone)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	r.curr.RowsData = remainRowsData
```
Comment (zz-jason) [inline distsql/stream.go]:
```go
if len(remainRowsData) == 0 {
    r.curr = nil
} else {
    r.curr.RowsData = remainRowsData
}
```

File: distsql/stream.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+// takeData ensures there're data is in current chunk. If no more data, return false.
+func (r *streamResult) takeData() error {
+	if r.curr != nil && len(r.curr.RowsData) > 0 {
+		return nil
+	}
+
+	tmp := new(tipb.Chunk)
+	finish, err := r.readDataFromResponse(r.resp, tmp)
+	if err != nil {
```
Comment (zz-jason) [inline distsql/stream.go]:
```go
if err != nil || finish {
    r.curr = nil
    return errors.Trace(err)
}
```

File: distsql/stream.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+	return r.scanCount
+}
+
+func (r *streamResult) Fetch(goctx.Context) {}
+
+func (r *streamResult) Next(goctx.Context) (PartialResult, error) {
+	var ret streamPartialResult
+	ret.rowLen = r.rowLen
+	finish, err := r.readDataFromResponse(r.resp, &ret.Chunk)
+	if err != nil {
```
Comment (zz-jason) [inline distsql/stream.go]: s/finish/finished/ and:
```go
if err != nil || finished {
    return nil, errors.Trace(err)
}
```

File: distsql/stream.go [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+		return false, errors.Trace(err)
+	}
+	if len(stream.OutputCounts) > 0 {
+		r.scanCount += stream.OutputCounts[0]
+	}
+	return false, nil
+}
+
+// takeData ensures there're data is in current chunk. If no more data, return false.
+func (r *streamResult) takeData() error {
```
Comment (zz-jason) [inline distsql/stream.go]: how about `s/takeData/readFromRespIfNecessary/` ?

File: distsql/stream.go (hunk 8 above)
Comment (tiancaiamao) [reply distsql/stream.go]: Finish branch and error branch do some different work, so it's better not to do the simplification.

**Comment by tiancaiamao** [comment]
PTAL @zimulala @zz-jason

**Comment by zz-jason** [comment]
LGTM

**Comment by shenli** [comment]
@coocood PTAL

File: distsql/stream.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+	finished, err := r.readDataFromResponse(r.resp, &ret.Chunk)
+	if err != nil || finished {
+		return nil, errors.Trace(err)
+	}
+	return &ret, nil
+}
+
+func (r *streamResult) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	for chk.NumRows() < r.ctx.GetSessionVars().MaxChunkSize {
```
Comment (coocood) [inline distsql/stream.go]: Make `MaxChunkSIze` a field of `streamResult` is more convenient.

Comment (tiancaiamao) [reply distsql/stream.go]: I make a temporary variable for it. @coocood

**Pull request merged**